core-client/transports/src/transports/ws.rs

```diff
@@ -75,67 +84,116 @@ where
 			queue: VecDeque::new(),
 		}
 	}
+
+	// Drains the internal buffer and attempts to forward as much of the items
+	// as possible to the underlying sink
+	fn try_empty_buffer(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), TSink::Error>> {
+		let this = Pin::into_inner(self);
+
+		match Pin::new(&mut this.sink).poll_ready(cx) {
+			Poll::Ready(value) => value?,
+			Poll::Pending => return Poll::Pending,
+		}
+
+		while let Some(item) = this.queue.pop_front() {
+			Pin::new(&mut this.sink).start_send(item)?;
+
+			if !this.queue.is_empty() {
+				match Pin::new(&mut this.sink).poll_ready(cx) {
+					Poll::Ready(value) => value?,
+					Poll::Pending => return Poll::Pending,
+				}
+			}
+		}
+
+		Poll::Ready(Ok(()))
+	}
 }
 
-impl<TSink, TStream, TError> Sink for WebsocketClient<TSink, TStream>
+// This mostly forwards to the underlying sink but also adds an unbounded queue
+// for when the underlying sink is incapable of receiving more items.
+// See https://docs.rs/futures-util/0.3.8/futures_util/sink/struct.Buffer.html
+// for the variant with a fixed-size buffer.
+impl<TSink, TStream> futures::Sink<String> for WebsocketClient<TSink, TStream>
 where
-	TSink: Sink<SinkItem = OwnedMessage, SinkError = TError>,
-	TStream: Stream<Item = OwnedMessage, Error = TError>,
-	TError: std::error::Error + Send + 'static,
+	TSink: futures::Sink<OwnedMessage, Error = RpcError> + Unpin,
+	TStream: futures::Stream<Item = Result<OwnedMessage, RpcError>> + Unpin,
 {
-	type SinkItem = String;
-	type SinkError = RpcError;
+	type Error = RpcError;
 
-	fn start_send(&mut self, request: Self::SinkItem) -> Result<AsyncSink<Self::SinkItem>, Self::SinkError> {
-		self.queue.push_back(OwnedMessage::Text(request));
-		Ok(AsyncSink::Ready)
+	fn start_send(mut self: Pin<&mut Self>, request: String) -> Result<(), Self::Error> {
+		let request = OwnedMessage::Text(request);
+
+		if self.queue.is_empty() {
+			let this = Pin::into_inner(self);
+			Pin::new(&mut this.sink).start_send(request)
+		} else {
+			self.queue.push_back(request);
+			Ok(())
+		}
 	}
 
-	fn poll_complete(&mut self) -> Result<Async<()>, Self::SinkError> {
-		loop {
-			match self.queue.pop_front() {
-				Some(request) => match self.sink.start_send(request) {
-					Ok(AsyncSink::Ready) => continue,
-					Ok(AsyncSink::NotReady(request)) => {
-						self.queue.push_front(request);
-						break;
-					}
-					Err(error) => return Err(RpcError::Other(Box::new(error))),
-				},
-				None => break,
-			}
+	fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+		let this = Pin::into_inner(self);
+
+		if this.queue.is_empty() {
+			return Pin::new(&mut this.sink).poll_ready(cx);
```

Comment:
Can you elaborate on this conditional? Why can't we just call `try_empty_buffer` instead?

Reply:
I was motivated by the code in `futures_util` - I like how it's explicit in the two related `start_send`/`poll_ready` that it simply delegates to the underlying sink if the buffer queue is empty. You're right, however, that this is equivalent to just calling `try_empty_buffer`. Do you want me to change it?

Reply:
Nah, let's keep it like this, was just curious if I'm missing something.